Fix side-effect of removal of equation_generation

garrigue opened the pull request:
In #13583, we removed the `equation_generation` logic, at it was unsound.
However, it means that `unify_gadt` can now fail because of non-injective type constructors.
This PR fixes that by first attempting to do normal unification in `unify_gadt` when there are type variables involved.

garrigue committed:
use backtracking in Ctype.unify_gadt

garrigue committed:
add new failure example

garrigue commented:
I also added a new failure example, which is more interesting, as one can see that the types are not normally unifiable.

gasche commented:
Outside the context of this fix, I think we can understand the new code as follows: we are trying to guarantee that turning a constructor into a GADT constructor will not break typing.

garrigue commented:
I did some measurements.
First using trunk (with #13583), calling `time make one DIR=tests/typing-gadts` 4 times.
```
5.504u 1.040s 0:06.96 93.9%	0+0k 0+0io 13710pf+0w
5.461u 0.976s 0:06.58 97.7%	0+0k 0+0io 13710pf+0w
5.460u 0.965s 0:06.56 97.8%	0+0k 0+0io 13710pf+0w
5.458u 0.969s 0:06.56 97.7%	0+0k 0+0io 13710pf+0w
```
then with this PR
```
5.512u 0.964s 0:06.61 97.8%	0+0k 0+0io 13710pf+0w
5.515u 0.976s 0:06.64 97.5%	0+0k 0+0io 13710pf+0w
5.525u 0.975s 0:06.64 97.7%	0+0k 0+0io 13710pf+0w
5.514u 0.967s 0:06.62 97.7%	0+0k 0+0io 13710pf+0w
```
So, there seems to be a 1% cost for GADT intensive code.

garrigue commented:
I also tried removing the `closed_type_expr` check, but there is a tiny further slowdown.
(About 0.2%)

Octachron commented:
Benchmarking on my GADT encoding that there are no [4x3 knight's tours](https://github.com/Octachron/typed-musings/blob/master/chess/tour.ml), there is an (expected) quite visible slowdown:

 
|           | Time (mean) |  σ        | min | max |   
|-------- | --------------------|---------| ---------|--------| 
| 5.2.0   |   691.3 ms |  15.5 ms |    668.6 ms | 736.2 ms|
| trunk  | 751.6 ms  | 11.5 ms | 727.3 ms |  793.5 ms|
| This PR | 1.361 s | 0.013 s | 1.330 s |  1.396 s |

gasche commented:
Yes, but (to help other people understand the discussion) as we discussed 2x is the worst case slowdown we can think of (for a code that spends all its time in GADT checking), so it's not actually so bad.

Octachron commented on typing/ctype.ml:
```diff
@@ -3259,15 +3259,28 @@ let unify uenv ty1 ty2 =
+    equated_types
+  in
+  let closed = closed_type_expr ty1 && closed_type_expr ty2 in
+  if closed then with_univar_pairs [] do_unify_gadt else
```
As an experiment selecting gadt unification as soon as `ty2` is closed restores trunk performance for my "knight's tour benchmark with no changes in the testsuite.

gasche replied:
Would `closed_type_expr ty1 || closed_type_expr ty2` work as well?

The way I understand only checking `ty2` is that we implicit assume that we are checking a more restricted "matching" problem in practice (only the left-hand-side has unification variables), but in practice `the unify_gadt`-using code in typecore.ml does not obviously appear to respect this asymmetric convention.

Octachron replied:
It doesn't work on the example in the testsuite (and the one reported in the previous PRs), since the GADT constructor has a closed type.

garrigue replied:
No, Florian's idea is to consider this as an asymmetric problem: since `unify_gadt` is only called for GADT constructors, and the type variables on the constructor instance side `ty1` are fresh, there is no problem in reifying them as long as they do not leak, and they won't if `ty2` is closed. If we do the same thing when  `ty1` only is closed, then we lose the example we are trying to fix.
The main concern is that we must now distinguish between the two arguments of `unify_gadt`. The calls in `solve_Ppat_construct` respect this asymmetry (with the twist that for some reason `unify_head_only` inverts its parameter order), but in `check_counter_example_pat`, all unifications call `unify_pat_types_refine ~refine:true`, so it is harder to ensure this asymmtery.  Yet, `check_partial` and `check_unused` are both called on `ty_arg_check`, which is a copy of `ty_arg'` in which type variables have been copied too, so they cannot escape either.
Conclusion, it seems that Florian's suggestion is correct, and further, calls from `check_counter_example_pat` do not need to bother about their free type variables.

gasche replied:
My intuition is that if the function is asymmetric, then its naming should say so, explain what is the desired use-case, and the callsites where it is not-obvious that the use-case is respected should be commented to explain why they are fine. In particular, you should rename `unify_head_only` and swap its parameter order.

Currently you are saying that the code is fine, but this is for a highly non-obvious reason that we will quickly forget, and we may break this invariant in the future. The result will be hard to detect because we don't have good coverage for these cases in our testsuite, and will lead to more confusion later.

garrigue replied:
Actually, `unify_head_only` is unrelated, since it only calls `unify_gadt` when typing counter-examples.
There is one single call-site outside of counter-examples, and it is `unify_res` inside `solve_Ppat_construct`.
I have already added a comment there, and there is a comment in ctype.mli for `unify_gadt`.
As I wrote below, the best way to clarify the code is to clarify the relation between `penv` and calls to `unify_gadt`, but it seems better done in a separate PR (I can of course do it in this one too).

garrigue committed:
optimize unify_gadt to only check for free variables in the expected type

garrigue commented:
I have added the suggested optimization.
Now the check is `penv.allow_recursive_equations || closed_type_expr ty2`.
I.e., if we are checking a counter-example, or if there are no free variables in the expected type, then no need to try normal unifcation.
I have added comments about the expected invariants.

To clarify the code, I plan to do the following refactoring in a separate PR:
* rename `allow_recursive_equations` to `force_refine` (or `counter_example` ?).
* in `Typecore`, remove `~refine` arguments, except in `unify_pat_types_return_equated_pairs`, which would become:
```ocaml
let unify_pat_types_return_equated_pairs ~refine loc penv ty ty' =
  try
    if refine || penv.force_refine then unify_gadt penv ty ty'
    else (unify !!penv ty ty'; nothing_equated)
  ...
```

garrigue committed:
add comment in unify_pat_types_refine

garrigue committed:
update comment in ctype.mli

OlivierNicole commented:
@Octachron Should this be on the 5.3 milestone?

garrigue commented:
#13599 is an alternative to this one. I.e., it does not fix the essence of #13598, but I'm starting to think that it is as good as this to fix the side effects.

Octachron commented:
I have sent a small commit that adds a label to `unify_gadt` to make the role of the two type expressions arguments more clearly asymmetrical in https://github.com/COCTI/ocaml/pull/16 .

My proposition for 5.3 would be to move forward with this PR as a safe fix, while possibly considering #13599 for later.

Octachron commented:
@garrigue , do you prefer that we merge this PR as it is? (I could send my sub PR as a refactorization PR)

garrigue commented:
I'm going to check that ASAP

garrigue commented:
Actually, if your read my comment above, the asymmetry is only in `unify_pat_types_return_equated_pairs`.
In particular all calls through `unify_pat_types_refine` are actually symmetrical (either normal unify, or both sides are reifiable).
Should I go forward and do the (small) refactoring suggested, as this makes all this clearer?

garrigue committed:
update result of examples

Octachron commented:
Sorry I missed (or forgot) your comment about this point, sorry about this !

Removing the duplication between the pattern_env and the refine argument outside of sounds like a nice refactorization but I agree that we can leave this improvement for a subsequent PR.

Octachron commented:
Cherry-picked on 5.3 as 32caed1e3, I have pulled along the refactoring from #13507 (in b73ea1e4) to reduce the distance between the trunk and 5.3 versions.
